lib/cn/cn_tree.c

```diff
@@ -123,6 +129,47 @@ cn_node_free(struct cn_tree_node *tn)
     }
 }
 
+static void
+cn_node_enqueue_subspill(struct subspill *ss, struct cn_tree_node *tn)
+{
+    struct list_head *p;
+    struct subspill *entry;
+
+    mutex_lock(&tn->tn_ss_lock);
+
+    /* Add ss at the right position in the node's mutation list. The list is sorted by
+     * sgen - smallest to largest.
+     */
+    list_for_each(p, &tn->tn_ss_list) {
+        entry = list_entry(p, typeof(*entry), ss_link);
+
+        if (ss->ss_sgen < entry->ss_sgen)
+            break;
+    }
+
+    list_add_tail(&ss->ss_link, p);
+    mutex_unlock(&tn->tn_ss_lock);
+}
+
+static struct subspill *
+cn_node_first_subspill(struct cn_tree_node *tn)
+{
+    struct subspill *entry;
+    bool found = false;
+
+    mutex_lock(&tn->tn_ss_lock);
+
+    entry = list_first_entry_or_null(&tn->tn_ss_list, typeof(*entry), ss_link);
+    if (entry && entry->ss_sgen == atomic_read(&tn->tn_sgen) + 1) {
+        list_del(&entry->ss_link);
+        found = true;
+    }
+
+    mutex_unlock(&tn->tn_ss_lock);
+
+    return found ? entry : NULL;
+}
```

Comment:
Seem like all the subspill functions should be like "cn_subspill_first()" or whatever, and take a struct subspill as it's only argument.  Then use the node from the subspill within the function.  Right?  That way you end up with a familiy of functions (cn_subspill_*) that clearly delineate the subspill API and make it easier to wrap your head around the operations provided.

Reply:
Yes, Fixed. Although this specific function operates on a cn_node, so I'll leave it as is - `cn_node_first_subspill()`

Reply:
Yeah, but it's primary concern is about subspill, so I'd change to make it part of the subspill family of functions...